StateMachineConfigurerAdapter doesn't find correct generics

Very similar to #306 but now for the `@EnableStateMachineFactory` part

If I have the following

```java
@EnableStateMachineFactory(name = "Foo")
FooStateMachineConfig extends StateMachineConfigurerAdapter<FooState, FooEvent>(){
}

@EnableStateMachineFactory(name = "Bar")
BarStateMachineConfig extends StateMachineConfigurerAdapter<BarState, BarEvent>(){
}
```

The following thing works

```java
@Autowired
@Qualifier("Foo")
private StateMachineFactory<FooState, FooEvent> stateMachineFactory;
```

But would be nice if the following would work as well
```java
@Autowired
private StateMachineFactory<FooState, FooEvent> stateMachineFactory;
```

Got it now in a small example and even with using the EnumStateMachineConfigurerAdapter

```java
@Component
public class SimpleTaskApp {

    @Autowired
    private StateMachineFactory<FooState, FooEvent> fooStateMachineFactory;
    @Autowired
    private StateMachineFactory<BarState, BarEvent> barStateMachineFactory;

    public void run() throws Exception {
        StateMachine<FooState, FooEvent> fooStateMachine = fooStateMachineFactory.getStateMachine();
        fooStateMachine.start();
        System.out.println(fooStateMachine.getState());

        StateMachine<BarState, BarEvent> barStateMachine = barStateMachineFactory.getStateMachine();
        barStateMachine.start();
        System.out.println(barStateMachine.getState());
    }

}


@Configuration
@EnableStateMachineFactory
public class FooStateMachineConfig extends EnumStateMachineConfigurerAdapter<FooStateMachineConfig.FooState, FooStateMachineConfig.FooEvent> {

    public static enum FooState {
        FOO_INIT, FOO_DONE
    }

    public static enum FooEvent {
        FOO_SUBMIT
    }

    @Override
    public void configure(StateMachineStateConfigurer<FooState, FooEvent> states) throws Exception {
        states.withStates()
                .initial(FooState.FOO_INIT)
                .states(EnumSet.allOf(FooState.class));
    }

    @Override
    public void configure(StateMachineTransitionConfigurer<FooState, FooEvent> transitions) throws Exception {
        transitions
                .withExternal()
                .source(FooState.FOO_INIT).target(FooState.FOO_DONE)
                .event(FooEvent.FOO_SUBMIT);
    }

}

@Configuration
@EnableStateMachineFactory
public class BarStateMachineConfig extends EnumStateMachineConfigurerAdapter<BarStateMachineConfig.BarState, BarStateMachineConfig.BarEvent> {

    public static enum BarState {
        BAR_INIT, BAR_DONE
    }

    public static enum BarEvent {
        BAR_SUBMIT
    }

    @Override
    public void configure(StateMachineStateConfigurer<BarState, BarEvent> states) throws Exception {
        states.withStates()
                .initial(BarState.BAR_INIT)
                .states(EnumSet.allOf(BarState.class));
    }

    @Override
    public void configure(StateMachineTransitionConfigurer<BarState, BarEvent> transitions) throws Exception {
        transitions
                .withExternal()
                .source(BarState.BAR_INIT).target(BarState.BAR_DONE)
                .event(BarEvent.BAR_SUBMIT);
    }

}
```

This runs but prints out
```
ObjectState [getIds()=[FOO_INIT], getClass()=class org.springframework.statemachine.state.ObjectState, hashCode()=1961176822, toString()=AbstractState [id=FOO_INIT, pseudoState=org.springframework.statemachine.state.DefaultPseudoState@6ed3ccb2, deferred=[], entryActions=[], exitActions=[], stateActions=[], regions=[], submachine=null]]
ObjectState [getIds()=[FOO_INIT], getClass()=class org.springframework.statemachine.state.ObjectState, hashCode()=1016550616, toString()=AbstractState [id=FOO_INIT, pseudoState=org.springframework.statemachine.state.DefaultPseudoState@3bf7ca37, deferred=[], entryActions=[], exitActions=[], stateActions=[], regions=[], submachine=null]]
```

But I would expect the second one to be `BAR_INIT`.

I'm guessing this is because the configs don't get a unique name automatically.
If I now add unique names to each of them

```java
@EnableStateMachineFactory(name = "foo")

@EnableStateMachineFactory(name = "bar")
```

I get

```
Caused by: org.springframework.beans.factory.NoUniqueBeanDefinitionException: No qualifying bean of type [org.springframework.statemachine.config.StateMachineFactory] is defined: expected single matching bean but found 2: bar,foo
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveNotUnique(DependencyDescriptor.java:172)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1106)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1056)
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:566)
	... 14 more

```

--- a/spring-statemachine-core/src/main/java/org/springframework/statemachine/config/configuration/StateMachineConfiguration.java
+++ b/spring-statemachine-core/src/main/java/org/springframework/statemachine/config/configuration/StateMachineConfiguration.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2015-2016 the original author or authors.
+ * Copyright 2015-2017 the original author or authors.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -121,7 +121,7 @@ public class StateMachineConfiguration<S, E> extends
 	private ResolvableType resolveFactoryObjectType(Class<?> enableStateMachineEnclosingClass) {
 		ResolvableType type = null;
 		try {
-			Class<?>[] generics = ResolvableType.forClass(enableStateMachineEnclosingClass).getSuperType().resolveGenerics();
+			Class<?>[] generics = ResolvableType.forClass(enableStateMachineEnclosingClass).as(StateMachineConfigurerAdapter.class).resolveGenerics();
 			if (generics != null && generics.length == 2) {
 				type = ResolvableType.forClassWithGenerics(StateMachine.class, generics);
 			}

--- a/spring-statemachine-core/src/main/java/org/springframework/statemachine/config/configuration/StateMachineFactoryConfiguration.java
+++ b/spring-statemachine-core/src/main/java/org/springframework/statemachine/config/configuration/StateMachineFactoryConfiguration.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2015-2016 the original author or authors.
+ * Copyright 2015-2017 the original author or authors.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -96,7 +96,7 @@ public class StateMachineFactoryConfiguration<S, E> extends
 	private ResolvableType resolveFactoryObjectType(Class<?> enableStateMachineEnclosingClass) {
 		ResolvableType type = null;
 		try {
-			Class<?>[] generics = ResolvableType.forClass(enableStateMachineEnclosingClass).getSuperType().resolveGenerics();
+			Class<?>[] generics = ResolvableType.forClass(enableStateMachineEnclosingClass).as(StateMachineConfigurerAdapter.class).resolveGenerics();
 			if (generics != null && generics.length == 2) {
 				type = ResolvableType.forClassWithGenerics(StateMachineFactory.class, generics);
 			}

--- a/spring-statemachine-core/src/test/java/org/springframework/statemachine/config/MachineTypedTests.java
+++ b/spring-statemachine-core/src/test/java/org/springframework/statemachine/config/MachineTypedTests.java
@@ -81,6 +81,28 @@ public class MachineTypedTests extends AbstractStateMachineTests {
 		assertThat(myBean4.someMachineFactory1, not(sameInstance(myBean4.someMachineFactory2)));
 	}
 
+	@Test
+	public void testAutowireMachineFactoryByNewBaseTypeNameNotMatches() {
+		context.register(Config5.class, Config6.class, MyBean5Config.class);
+		context.refresh();
+
+		MyBean5 myBean5 = context.getBean(MyBean5.class);
+		assertThat(myBean5.someMachineFactory3, notNullValue());
+		assertThat(myBean5.someMachineFactory4, notNullValue());
+		assertThat(myBean5.someMachineFactory3, not(sameInstance(myBean5.someMachineFactory4)));
+	}
+
+	@Test
+	public void testAutowireMachineByNewBaseTypeNameNotMatches() {
+		context.register(Config7.class, Config8.class, MyBean6Config.class);
+		context.refresh();
+
+		MyBean6 myBean6 = context.getBean(MyBean6.class);
+		assertThat(myBean6.someMachine3, notNullValue());
+		assertThat(myBean6.someMachine4, notNullValue());
+		assertThat(myBean6.someMachine3, not(sameInstance(myBean6.someMachine4)));
+	}
+
 	@Configuration
 	public static class MyBean1Config {
 
@@ -114,6 +136,24 @@ public class MachineTypedTests extends AbstractStateMachineTests {
 		@Bean
 		public MyBean4 myBean4() {
 			return new MyBean4();
+		}
+	}
+
+	@Configuration
+	public static class MyBean5Config {
+
+		@Bean
+		public MyBean5 myBean5() {
+			return new MyBean5();
+		}
+	}
+
+	@Configuration
+	public static class MyBean6Config {
+
+		@Bean
+		public MyBean6 myBean6() {
+			return new MyBean6();
 		}
 	}
 
@@ -151,6 +191,24 @@ public class MachineTypedTests extends AbstractStateMachineTests {
 
 		@Autowired
 		StateMachineFactory<MyTestStates2, MyTestEvents2> someMachineFactory2;
+	}
+
+	public static class MyBean5 {
+
+		@Autowired
+		StateMachineFactory<MyTestStates1, MyTestEvents1> someMachineFactory3;
+
+		@Autowired
+		StateMachineFactory<MyTestStates2, MyTestEvents2> someMachineFactory4;
+	}
+
+	public static class MyBean6 {
+
+		@Autowired
+		StateMachine<MyTestStates1, MyTestEvents1> someMachine3;
+
+		@Autowired
+		StateMachine<MyTestStates2, MyTestEvents2> someMachine4;
 	}
 
 	@Configuration
@@ -235,6 +293,96 @@ public class MachineTypedTests extends AbstractStateMachineTests {
 					.source(MyTestStates2.S1).target(MyTestStates2.S2)
 					.event(MyTestEvents2.E1);
 		}
+	}
+
+	@Configuration
+	@EnableStateMachineFactory(name = "machinefactory3")
+	public static class Config5 extends MyNewBaseAdapter1 {
+
+		@Override
+		public void configure(StateMachineStateConfigurer<MyTestStates1, MyTestEvents1> states) throws Exception {
+			states
+				.withStates()
+					.initial(MyTestStates1.S1)
+					.state(MyTestStates1.S2);
+		}
+
+		@Override
+		public void configure(StateMachineTransitionConfigurer<MyTestStates1, MyTestEvents1> transitions) throws Exception {
+			transitions
+				.withExternal()
+					.source(MyTestStates1.S1).target(MyTestStates1.S2)
+					.event(MyTestEvents1.E1);
+		}
+	}
+
+	@Configuration
+	@EnableStateMachineFactory(name = "machinefactory4")
+	public static class Config6 extends MyNewBaseAdapter2 {
+
+		@Override
+		public void configure(StateMachineStateConfigurer<MyTestStates2, MyTestEvents2> states) throws Exception {
+			states
+				.withStates()
+					.initial(MyTestStates2.S1)
+					.state(MyTestStates2.S2);
+		}
+
+		@Override
+		public void configure(StateMachineTransitionConfigurer<MyTestStates2, MyTestEvents2> transitions) throws Exception {
+			transitions
+				.withExternal()
+					.source(MyTestStates2.S1).target(MyTestStates2.S2)
+					.event(MyTestEvents2.E1);
+		}
+	}
+
+	@Configuration
+	@EnableStateMachine(name = "machine3")
+	public static class Config7 extends MyNewBaseAdapter1 {
+
+		@Override
+		public void configure(StateMachineStateConfigurer<MyTestStates1, MyTestEvents1> states) throws Exception {
+			states
+				.withStates()
+					.initial(MyTestStates1.S1)
+					.state(MyTestStates1.S2);
+		}
+
+		@Override
+		public void configure(StateMachineTransitionConfigurer<MyTestStates1, MyTestEvents1> transitions) throws Exception {
+			transitions
+				.withExternal()
+					.source(MyTestStates1.S1).target(MyTestStates1.S2)
+					.event(MyTestEvents1.E1);
+		}
+	}
+
+	@Configuration
+	@EnableStateMachine(name = "machine4")
+	public static class Config8 extends MyNewBaseAdapter2 {
+
+		@Override
+		public void configure(StateMachineStateConfigurer<MyTestStates2, MyTestEvents2> states) throws Exception {
+			states
+				.withStates()
+					.initial(MyTestStates2.S1)
+					.state(MyTestStates2.S2);
+		}
+
+		@Override
+		public void configure(StateMachineTransitionConfigurer<MyTestStates2, MyTestEvents2> transitions) throws Exception {
+			transitions
+				.withExternal()
+					.source(MyTestStates2.S1).target(MyTestStates2.S2)
+					.event(MyTestEvents2.E1);
+		}
+	}
+
+	public static class MyNewBaseAdapter1 extends EnumStateMachineConfigurerAdapter<MyTestStates1, MyTestEvents1> {
+	}
+
+	public static class MyNewBaseAdapter2 extends EnumStateMachineConfigurerAdapter<MyTestStates2, MyTestEvents2> {
 	}
 
 	public enum MyTestStates1 {

